runtime/vm/VMSnapshotImpl.cpp

```diff
@@ -1269,65 +1531,344 @@ resetVMThreadState(J9JavaVM *vm, J9VMThread *restoreThread)
+		if (false == rc) {
+			goto done;
+		}
+
+		rc = restoreSystemMonitor(&monitorData, &monitorCount, _vm->verboseStateMutex, FIXUPREFVM_VERBOSE_STATE_MUTEX);
+		if (false == rc) {
+			goto done;
+		}
+
+		rc = restoreSystemMonitor(&monitorData,  &monitorCount, _vm->jclCacheMutex, FIXUPREFVM_JCL_CACHE_MUTEX);
+		if (false == rc) {
+			goto done;
+		}
+
+		rc = restoreSystemMonitor(&monitorData, &monitorCount, _vm->constantDynamicMutex, FIXUPREFVM_CONSTANT_DYNAMIC_MUTEX); /* Java 11 */
+		if (false == rc) {
+			goto done;
+		}
+
+		/* Restore object monitors. This should be done one thread at a time in the same order in which they were saved. */
+		threadCursor = _vm->mainThread;
+		do {
+			infoLen = vmFuncs->getOwnedObjectMonitors(threadCursor, threadCursor, NULL, 0);
+			if (infoLen > 0) {
+				PORT_ACCESS_FROM_PORT(_portLibrary);
+
+				J9ObjectMonitorInfo *info = (J9ObjectMonitorInfo*) j9mem_allocate_memory(infoLen * sizeof(J9ObjectMonitorInfo), J9MEM_CATEGORY_CLASSES);
+				if (NULL == info) {
+					goto done;
+				}
+
+				rc = vmFuncs->getOwnedObjectMonitors(threadCursor, threadCursor, info, infoLen);
+				if (false == rc) {
+					j9mem_free_memory(info);
+					goto done;
+				}
+
+				/* process object monitors - should be in the same order as during snapshot */
+				for (IDATA i = 0; i < infoLen; i++) {
+					/* objects should be in the same order as on restore. */
+					rc = restoreAcquiredObjectMonitor(&monitorData, &monitorCount, info[i].object);
+					if (false == rc) {
+						j9mem_free_memory(info);
+						goto done;
+					}
+				}
+
+				j9mem_free_memory(info);
+
+			}
+
+			threadCursor = threadCursor->linkNext;
+		} while (threadCursor != _vm->mainThread);
+
+		if (monitorCount > 0) {
+			printf("ERROR: somehow not all monitors were restored\n");
+			rc = false;
+		}
 	}
 
-#ifdef J9VM_IVE_RAW_BUILD /* J9VM_IVE_RAW_BUILD is not enabled by default */
-	{
-		j9object_t unicodeChars = J9VMJAVALANGTHREAD_NAME(restoreThread, restoreThread->threadObject);
-		if (NULL != unicodeChars) {
-			threadName = copyStringToUTF8WithMemAlloc(restoreThread, unicodeChars, J9_STR_NULL_TERMINATE_RESULT, "", 0, NULL, 0, NULL);
+done:
+	return rc;
+}
+
+/**
+ * If vmThread should be waiting on a system monitor
+ * attempt to acquire it here.
+ * @param vmThread
+ */
+static void
+restoreSystemMonitorWaiting(J9VMThread* vmThread) {
+	UDATA fixupId = vmThread->waitingMonitorFixupId;
+	if (0 != fixupId) {
+		J9JavaVM* vm = vmThread->javaVM;
+		switch(fixupId) {
+		case FIXUPREFVM_UNSAFE_MEMORY_TRACKING_MUTEX:
+			omrthread_monitor_enter(vm->unsafeMemoryTrackingMutex);
+		case FIXUPREFVM_VERBOSE_STATE_MUTEX:
+			omrthread_monitor_enter(vm->verboseStateMutex);
+		case FIXUPREFVM_JCL_CACHE_MUTEX:
+			omrthread_monitor_enter(vm->jclCacheMutex);
+		case FIXUPREFVM_CONSTANT_DYNAMIC_MUTEX:
+			omrthread_monitor_enter(vm->constantDynamicMutex);
```

Comment:
If the fallthru here is deliberate, then it should be commented.  Otherwise, break statements should be added